scene/gui/button.cpp

```diff
@@ -213,146 +262,160 @@ void Button::_notification(int p_what) {
+			const Size2 custom_element_size = drawable_size_remained;
 
-				// Fix vertical size.
-				if (vertical_icon_alignment != VERTICAL_ALIGNMENT_CENTER) {
-					valign -= text_buf->get_size().height;
-				}
+			// Draw the icon.
+			if (_icon.is_valid()) {
+				Size2 icon_size;
+				Point2 icon_ofs;
 
-				float icon_ofs_region = 0.0;
-				Point2 style_offset;
-				if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_LEFT) {
-					style_offset.x = style->get_margin(SIDE_LEFT);
-					if (_internal_margin[SIDE_LEFT] > 0) {
-						icon_ofs_region = _internal_margin[SIDE_LEFT] + theme_cache.h_separation;
-					}
-				} else if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_CENTER) {
-					style_offset.x = 0.0;
-				} else if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_RIGHT) {
-					style_offset.x = -style->get_margin(SIDE_RIGHT);
-					if (_internal_margin[SIDE_RIGHT] > 0) {
-						icon_ofs_region = -_internal_margin[SIDE_RIGHT] - theme_cache.h_separation;
+				{ // Calculate the drawing size of the icon.
+					icon_size = _icon->get_size();
+
+					if (expand_icon) {
+						const Size2 text_buf_size = text_buf->get_size();
+						Size2 _size = custom_element_size;
+						if (!is_clipped && icon_align_rtl_checked != HORIZONTAL_ALIGNMENT_CENTER) {
+							_size.width -= text_buf_size.width;
+						}
+						if (vertical_icon_alignment != VERTICAL_ALIGNMENT_CENTER) {
+							_size.height -= text_buf_size.height;
+						}
+
+						float icon_width = icon_size.width * _size.height / icon_size.height;
+						float icon_height = _size.height;
+
+						if (icon_width > _size.width) {
+							icon_width = _size.width;
+							icon_height = icon_size.height * icon_width / icon_size.width;
+						}
+
+						icon_size = Size2(icon_width, icon_height);
 					}
+					icon_size = _fit_icon_size(icon_size);
 				}
-				style_offset.y = style->get_margin(SIDE_TOP);
-
-				if (expand_icon) {
-					Size2 _size = get_size() - style->get_offset() * 2;
-					int icon_text_separation = text.is_empty() ? 0 : theme_cache.h_separation;
-					_size.width -= icon_text_separation + icon_ofs_region;
-					if (!is_clipped && icon_align_rtl_checked != HORIZONTAL_ALIGNMENT_CENTER) {
-						_size.width -= text_buf->get_size().width;
-					}
-					if (vertical_icon_alignment != VERTICAL_ALIGNMENT_CENTER) {
-						_size.height -= text_buf->get_size().height;
-					}
-					float icon_width = _icon->get_width() * _size.height / _icon->get_height();
-					float icon_height = _size.height;
 
-					if (icon_width > _size.width) {
-						icon_width = _size.width;
-						icon_height = _icon->get_height() * icon_width / _icon->get_width();
+				{ // Calculate the drawing position of the icon.
+					icon_ofs = (custom_element_size - icon_size) / 2.0f;
+
+					switch (icon_align_rtl_checked) {
+						case HORIZONTAL_ALIGNMENT_FILL:
+						case HORIZONTAL_ALIGNMENT_LEFT: {
+							icon_ofs.x = style_margin_left;
+							if (internal_margin > 0.0f) {
+								icon_ofs.x += internal_margin;
```

Comment:
So I guess it's for this bit, and for the text below. This confirms to me that this implementation is lacking. We need to account for both sides, and we cannot use the sign for that.

Reply:
Although the size of `_internal_margin` is 4, only 2 elements are valid. Compared with `_internal_margin`, `internal_margin` also contains possible `h_separation`. 

`internal_margin > 0.0f` is equivalent to `_internal_margin[SIDE_LEFT] > _internal_margin[SIDE_RIGHT]`, while `internal_margin < 0.0f` is equivalent to `_internal_margin[SIDE_LEFT] < _internal_margin[SIDE_RIGHT]`

Maybe the second comparison and the positive values will make them clearer.